pkg/oauth/tokenrequest/basicauth.go

```diff
@@ -0,0 +1,160 @@
+package tokenrequest
+
+import (
+	"encoding/base64"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"regexp"
+	"strings"
+
+	"k8s.io/klog/v2"
+)
+
+// BasicAuthNoUsernameError when basic authentication challenge handling was attempted
+// but the required username was not provided from the command line options
+type BasicAuthNoUsernameError struct{}
```

Comment:
why does this needs to be fully qualified type?

Reply:
so that you are able to distinguish the error outside of the library

Reply:
ack, just a suggestion to consider other options as well, see https://dave.cheney.net/2016/04/27/dont-just-check-errors-handle-them-gracefully. The "Assert errors for behaviour, not type" is a good one.

this seems fine for now though.